src/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs

```diff
@@ -426,13 +426,68 @@ internal void AddInterfaceImpl(Type iface)
                 _assembly.EnsureTypeIsVisible(iface);
 
                 _tb.AddInterfaceImplementation(iface);
+
+                // AccessorMethods -> Metadata mappings.
+                var propertyMap = new Dictionary<MethodInfo, PropertyAccessorInfo>();
```

Comment:
There's a hidden hazard here in using MethodInfo as a dictionary key. The default Equality test for MethodInfo is reference equality, not "semantic equality" and the framework doesn't give any strict guarantees as to whether you get the same MethodInfo back each time (the actual outcome is determined by the framework's internal caching heuristics, not by any api rule.)

We don't have a great guidance story around this gotcha but in this case, I think the most pragmatic approach is:

- Pass in an IEqualityComparer<MethodInfo> to the dictionary, that you implement in a private class.

- Implement the Compare() method as:

  - Compare the method's DeclaringType using Equals().
  - Compare the method's MetadataToken value.